quint/src/effects/builtinSignatures.ts

```diff
@@ -12,175 +12,241 @@
  * @module
  */
 
-import { EffectScheme, Signature, effectNames } from './base'
+import { ComponentKind, Effect, EffectComponent, EffectScheme, Signature, Variables, effectNames, toScheme } from './base'
 import { parseEffectOrThrow } from './parser'
-import { range, times, zip } from 'lodash'
+import { range, times } from 'lodash'
 
 export function getSignatures(): Map<string, Signature> {
   return new Map<string, Signature>(fixedAritySignatures.concat(multipleAritySignatures))
 }
 
-const literals = ['Nat', 'Int', 'Bool'].map(name => ({ name, effect: 'Pure' }))
-
 function parseAndQuantify(effectString: string): EffectScheme {
   const e = parseEffectOrThrow(effectString)
   return { effect: e, ...effectNames(e) }
 }
 
+const prefixByComponentKind = new Map<ComponentKind, string>([
+  ['read', 'r'],
+  ['temporal', 't'],
+  ['update', 'u'],
+])
+
+/**
+ * Builds a signature that propagates components of the given kinds
+ *
+ * Example: propagateComponents(['read', 'temporal'])(2) results in
+ * (Read[r1] & Temporal[t1], Read[r2] & Temporal[t2]) => Read[r1, r2] & Temporal[t1, t2]
+ *
+ * @param kinds the kinds of components to propagate
+ * @returns an arrow function taking arity and returning the effect for that arity
+ */
+function propagateComponents(kinds: ComponentKind[]): ((arity: number) => EffectScheme) {
+  return (arity: number) => {
+    const params: Effect[] = times(arity, i => {
+      const components: EffectComponent[] = kinds.map(kind => {
+        return { kind: kind, variables: { kind: 'quantified', name: `${prefixByComponentKind.get(kind)!}${i + 1}` } }
+      })
+
+      return { kind: 'concrete', components: components }
+    })
+
+    const resultComponents: EffectComponent[] = kinds.map(kind => {
+      const names = times(arity, i => `${prefixByComponentKind.get(kind)!}${i + 1}`)
+      const variables: Variables[] = names.map(name => ({ kind: 'quantified', name }))
+      return { kind: kind, variables: { kind: 'union', variables } }
+    })
+
+    const result: Effect = { kind: 'concrete', components: resultComponents }
+    const effect: Effect = { kind: 'arrow', params: params, result: result }
+
+    return { effect, ...effectNames(effect) }
+  }
+}
+
+const p = parseAndQuantify
```

Comment:
This abbrev. seems ex. shrt.; poss. to mk. l.? :stuck_out_tongue: 

The abbreviation seems too cryptic to me given the specificity of the value and how big its scope is. Do you think something a bit more descriptive could be found?

Reply:
I intentionally used the shortest possibility to clean out the rest of the file since this is used a lot, and frequently with long strings that take a lot of line space. I can probably achieve the same with indentation, I'll try.